csds-client/client/util.go

```diff
@@ -0,0 +1,201 @@
+func (r *TypeResolver) FindMessageByName(message protoreflect.FullName) (protoreflect.MessageType, error) {
+	return nil, protoregistry.NotFound
+}
+
+// link the message type url to the specific message type
+func (r *TypeResolver) FindMessageByURL(url string) (protoreflect.MessageType, error) {
+	if url == "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager" {
+		httpConnectionManager := envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager{}
+		return httpConnectionManager.ProtoReflect().Type(), nil
+	}
+	if url == "type.googleapis.com/envoy.api.v2.Cluster" {
+		cluster := envoy_api_v2.Cluster{}
+		return cluster.ProtoReflect().Type(), nil
+	}
+	if url == "type.googleapis.com/envoy.api.v2.Listener" {
+		listener := envoy_api_v2.Listener{}
+		return listener.ProtoReflect().Type(), nil
+	}
+	if url == "type.googleapis.com/envoy.config.filter.http.router.v2.Router" {
+		router := envoy_config_filter_http_router_v2.Router{}
+		return router.ProtoReflect().Type(), nil
+	}
+	if url == "type.googleapis.com/envoy.api.v2.RouteConfiguration" {
+		routeConfiguration := envoy_api_v2.RouteConfiguration{}
+		return routeConfiguration.ProtoReflect().Type(), nil
+	}
+	return nil, protoregistry.NotFound
+}
+
+func (r *TypeResolver) FindExtensionByName(field protoreflect.FullName) (protoreflect.ExtensionType, error) {
+	return nil, protoregistry.NotFound
+}
+
+func (r *TypeResolver) FindExtensionByNumber(message protoreflect.FullName, field protoreflect.FieldNumber) (protoreflect.ExtensionType, error) {
+	return nil, protoregistry.NotFound
+}
+
+// post process response
+func ParseResponse(response *envoy_service_status_v2.ClientStatusResponse, fileName string) {
+	fmt.Printf("%-50s %-30s %-30s \n", "Client ID", "xDS stream type", "Config")
+	for _, config := range response.Config {
+		id := config.Node.GetId()
+		metadata := config.Node.GetMetadata().AsMap()
+		xdsType := metadata["TRAFFIC_DIRECTOR_XDS_STREAM_TYPE"]
+		if xdsType == nil {
+			xdsType = ""
+		}
+
+		var configFile string
+		if config.GetXdsConfig() == nil {
+			configFile = "N/A"
+			fmt.Printf("%-50s %-30s %-30s \n", id, xdsType, configFile)
+		} else {
+			// parse response to json
+			// format the json and resolve google.protobuf.Any types
+			m := protojson.MarshalOptions{Multiline: true, Indent: "  ", Resolver: &TypeResolver{}}
+			out, err := m.Marshal(response)
+			if err != nil {
+				fmt.Printf("%v", err)
+			}
+			if fileName == "" {
+				// output the configuration to stdout by default
+				fmt.Printf("%-50s %-30s %-30s \n", id, xdsType, configFile)
+				fmt.Println(string(out))
+			} else {
+				// write the configuration to the file
+				configFile = fileName
+				f, err := os.Create(configFile)
+				if err != nil {
+					fmt.Println(err)
+				}
+				defer f.Close()
+				_, err = f.Write(out)
+				if err != nil {
+					fmt.Printf("%v", err)
+				}
+				fmt.Printf("%-50s %-30s %-30s \n", id, xdsType, configFile)
```

Comment:
Can this summary also include the Status of this Node's config?  See https://www.envoyproxy.io/docs/envoy/latest/api-v2/service/status/v2/csds.proto#envoy-api-enum-service-status-v2-configstatus

Reply:
For one Node, each xDS config will includes a ConfigStatus. So there would be multiple statuses for one node. Printing all of them out may be a little hard to read?

Reply:
Or maybe I can print all the status with xDS, for example
```
Client ID                      xDS stream type                Config             Config Status                       
<client_id>                    ADS                            myconfig.json      LDS  SYNCED
                                                                                 RDS  SYNCED
                                                                                 CDS  STALE
```

Reply:
ConfigStatus has already been included in the output summary like
```
Client ID                      xDS stream type                  Config Status
<client_id>                    ADS                              LDS  SYNCED
                                                                RDS  SYNCED
                                                                CDS  STALE
```
This feature has been updated in [README](https://github.com/Rainton/envoy-tools/blob/feature-branch/csds-client/README.md)